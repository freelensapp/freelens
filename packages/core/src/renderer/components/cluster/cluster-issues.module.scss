/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

 .ClusterIssues {
  min-height: 350px;
  position: relative;
  grid-area: issues;
  padding: calc(var(--margin) * 2) 0;
  background: var(--contentColor);

  &:global(.OnlyClusterIssues) {
    grid-row: row1-start / row2-end;
  }

  @media (max-width: 1150px) {
    grid-area: unset;

    &:global(.OnlyClusterIssues) {
      grid-row: auto;
    }
  }

  .SubHeader {
    padding-top: 0;
    padding-bottom: 0;

    :global(.Icon) {
      font-size: 130%;
      color: var(--colorError);
    }
  }

  .message {
    white-space: nowrap;
    text-overflow: ellipsis;
    flex-grow: 3;
  }

  .object {
    white-space: nowrap;
    text-overflow: ellipsis;
    flex-grow: 2;
  }

   .noIssues {
     .Icon {
       color: white;
     }

     .title {
       font-size: large;
       color: var(--textColorAccent);
       font-weight: bold;
     }
   }
}
