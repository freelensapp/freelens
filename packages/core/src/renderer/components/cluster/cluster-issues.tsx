/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import styles from "./cluster-issues.module.scss";

import { Icon } from "@freelensapp/icon";
import { Spinner } from "@freelensapp/spinner";
import { cssNames, prevDefault } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import type { IComputedValue } from "mobx";
import { computed, makeObservable } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { ApiManager } from "../../../common/k8s-api/api-manager";
import apiManagerInjectable from "../../../common/k8s-api/api-manager/manager.injectable";
import type { PageParam } from "../../navigation/page-param";
import activeThemeInjectable from "../../themes/active.injectable";
import type { LensTheme } from "../../themes/lens-theme";
import type { EventStore } from "../events/store";
import eventStoreInjectable from "../events/store.injectable";
import kubeSelectedUrlParamInjectable from "../kube-detail-params/kube-selected-url.injectable";
import type { ToggleKubeDetailsPane } from "../kube-detail-params/toggle-details.injectable";
import toggleKubeDetailsPaneInjectable from "../kube-detail-params/toggle-details.injectable";
import { KubeObjectAge } from "../kube-object/age";
import { SubHeader } from "../layout/sub-header";
import type { NodeStore } from "../nodes/store";
import nodeStoreInjectable from "../nodes/store.injectable";
import { Table, TableCell, TableHead, TableRow } from "../table";

export interface ClusterIssuesProps {
  className?: string;
}

interface Warning {
  getId: () => string;
  getName: () => string;
  kind: string;
  message: string | undefined;
  selfLink: string;
  renderAge: () => React.ReactElement;
  ageMs: number;
}

enum sortBy {
  type = "type",
  object = "object",
  age = "age",
}

interface Dependencies {
  activeTheme: IComputedValue<LensTheme>;
  nodeStore: NodeStore;
  eventStore: EventStore;
  apiManager: ApiManager;
  kubeSelectedUrlParam: PageParam<string>;
  toggleKubeDetailsPane: ToggleKubeDetailsPane;
}

@observer
class NonInjectedClusterIssues extends React.Component<ClusterIssuesProps & Dependencies> {
  constructor(props: ClusterIssuesProps & Dependencies) {
    super(props);
    makeObservable(this);
  }

  @computed get warnings(): Warning[] {
    return [
      ...this.props.nodeStore.items.flatMap((node) =>
        node.getWarningConditions().map(({ message }) => ({
          selfLink: node.selfLink,
          getId: () => node.getId(),
          getName: () => node.getName(),
          kind: node.kind,
          message,
          renderAge: () => <KubeObjectAge key="age" object={node} />,
          ageMs: -node.getCreationTimestamp(),
        })),
      ),
      ...this.props.eventStore.getWarnings().map((warning) => ({
        getId: () => warning.involvedObject.uid,
        getName: () => warning.involvedObject.name,
        renderAge: () => <KubeObjectAge key="age" object={warning} />,
        ageMs: -warning.getCreationTimestamp(),
        message: warning.message,
        kind: warning.kind,
        selfLink: this.props.apiManager.lookupApiLink(warning.involvedObject, warning),
      })),
    ];
  }

  getTableRow = (uid: string) => {
    const { warnings } = this;
    const { kubeSelectedUrlParam, toggleKubeDetailsPane: toggleDetails } = this.props;
    const warning = warnings.find((warn) => warn.getId() == uid);

    if (!warning) {
      return undefined;
    }

    const { getId, getName, message, kind, selfLink, renderAge } = warning;

    return (
      <TableRow
        key={getId()}
        sortItem={warning}
        selected={selfLink === kubeSelectedUrlParam.get()}
        onClick={prevDefault(() => toggleDetails(selfLink))}
      >
        <TableCell className={styles.message}>{message ?? "<unknown>"}</TableCell>
        <TableCell className={styles.object}>{getName()}</TableCell>
        <TableCell className="kind">{kind}</TableCell>
        <TableCell className="age">{renderAge()}</TableCell>
      </TableRow>
    );
  };

  renderContent() {
    const { warnings } = this;

    if (!this.props.eventStore.isLoaded) {
      return <Spinner center />;
    }

    if (!warnings.length) {
      return (
        <div className={cssNames(styles.noIssues, "flex column box grow gaps align-center justify-center")}>
          <Icon className={styles.Icon} material="check" big sticker />
          <p className={styles.title}>No issues found</p>
          <p>Everything is fine in the Cluster</p>
        </div>
      );
    }

    return (
      <>
        <SubHeader className={styles.SubHeader}>
          <Icon material="error_outline" />
          {` Warnings: ${warnings.length}`}
        </SubHeader>
        <Table
          tableId="cluster_issues"
          items={warnings}
          virtual
          selectable
          sortable={{
            [sortBy.type]: (warning) => warning.kind,
            [sortBy.object]: (warning) => warning.getName(),
            [sortBy.age]: (warning) => warning.ageMs,
          }}
          sortByDefault={{ sortBy: sortBy.object, orderBy: "asc" }}
          sortSyncWithUrl={false}
          getTableRow={this.getTableRow}
          className={cssNames("box grow", this.props.activeTheme.get().type)}
        >
          <TableHead nowrap>
            <TableCell className="message">Message</TableCell>
            <TableCell className="object" sortBy={sortBy.object}>
              Object
            </TableCell>
            <TableCell className="kind" sortBy={sortBy.type}>
              Type
            </TableCell>
            <TableCell className="timestamp" sortBy={sortBy.age}>
              Age
            </TableCell>
          </TableHead>
        </Table>
      </>
    );
  }

  render() {
    return (
      <div className={cssNames(styles.ClusterIssues, "flex column", this.props.className)}>{this.renderContent()}</div>
    );
  }
}

export const ClusterIssues = withInjectables<Dependencies, ClusterIssuesProps>(NonInjectedClusterIssues, {
  getProps: (di, props) => ({
    ...props,
    activeTheme: di.inject(activeThemeInjectable),
    apiManager: di.inject(apiManagerInjectable),
    eventStore: di.inject(eventStoreInjectable),
    nodeStore: di.inject(nodeStoreInjectable),
    kubeSelectedUrlParam: di.inject(kubeSelectedUrlParamInjectable),
    toggleKubeDetailsPane: di.inject(toggleKubeDetailsPaneInjectable),
  }),
});
