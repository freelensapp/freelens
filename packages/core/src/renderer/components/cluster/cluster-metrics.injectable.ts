/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectable } from "@ogre-tools/injectable";
import { asyncComputed } from "@ogre-tools/injectable-react";
import { now } from "mobx-utils";
import type { ClusterMetricData } from "../../../common/k8s-api/endpoints/metrics.api/request-cluster-metrics-by-node-names.injectable";
import requestClusterMetricsByNodeNamesInjectable from "../../../common/k8s-api/endpoints/metrics.api/request-cluster-metrics-by-node-names.injectable";
import selectedNodeRoleForMetricsInjectable from "./overview/selected-node-role-for-metrics.injectable";

const everyMinute = 60 * 1000;

const clusterOverviewMetricsInjectable = getInjectable({
  id: "cluster-overview-metrics",
  instantiate: (di) => {
    const requestClusterMetricsByNodeNames = di.inject(requestClusterMetricsByNodeNamesInjectable);
    const selectedNodeRoleForMetrics = di.inject(selectedNodeRoleForMetricsInjectable);

    return asyncComputed<ClusterMetricData | undefined>({
      getValueFromObservedPromise: async () => {
        now(everyMinute);

        const nodeNames = selectedNodeRoleForMetrics.nodes.get().map((node) => node.getName());

        return requestClusterMetricsByNodeNames(nodeNames);
      },
      betweenUpdates: "show-latest-value",
    });
  },
});

export default clusterOverviewMetricsInjectable;
