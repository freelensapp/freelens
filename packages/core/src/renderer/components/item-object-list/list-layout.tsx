/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./item-list-layout.scss";

import type { ItemObject, TableCellProps } from "@freelensapp/list-layout";
import type { IClassName, SingleOrMany, StrictReactNode } from "@freelensapp/utilities";
import { cssNames, noop } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import autoBindReact from "auto-bind/react";
import groupBy from "lodash/groupBy";
import type { IComputedValue } from "mobx";
import { computed, makeObservable, untracked } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { Primitive } from "type-fest";
import selectedFilterNamespacesInjectable from "../../../common/k8s-api/selected-filter-namespaces.injectable";
import type { SubscribableStore } from "../../kube-watch-api/kube-watch-api";
import type { StorageLayer } from "../../utils/storage-helper";
import type { AddRemoveButtonsProps } from "../add-remove-buttons";
import type { ConfirmDialogParams } from "../confirm-dialog";
import type { SearchInputUrlProps } from "../input";
import type { TableProps, TableRowProps, TableSortCallbacks } from "../table";
import { ItemListLayoutContent } from "./content";
import { ItemListLayoutFilters } from "./filters";
import { ItemListLayoutHeader } from "./header";
import { PageFiltersList } from "./page-filters/list";
import type { PageFiltersStore } from "./page-filters/store";
import { FilterType } from "./page-filters/store";
import pageFiltersStoreInjectable from "./page-filters/store.injectable";
import itemListLayoutStorageInjectable from "./storage.injectable";

export type SearchFilter<I extends ItemObject> = (item: I) => SingleOrMany<string | number | undefined | null>;
export type SearchFilters<I extends ItemObject> = Record<string, SearchFilter<I>>;
export type ItemsFilter<I extends ItemObject> = (items: I[]) => I[];
export type ItemsFilters<I extends ItemObject> = Record<string, ItemsFilter<I>>;

export interface HeaderPlaceholders {
  title?: StrictReactNode;
  searchProps?: SearchInputUrlProps;
  filters?: StrictReactNode;
  info?: StrictReactNode;
}

function normalizeText(value: Primitive) {
  return String(value).toLowerCase();
}

export type ItemListStore<I extends ItemObject, PreLoadStores extends boolean> = {
  readonly isLoaded: boolean;
  readonly failedLoading: boolean;
  getTotalCount: () => number;
  isSelected: (item: I) => boolean;
  toggleSelection: (item: I) => void;
  isSelectedAll: (items: I[]) => boolean;
  toggleSelectionAll: (enabledItems: I[]) => void;
  pickOnlySelected: (items: I[]) => I[];
} & (
  | {
      removeItems: (selectedItems: I[]) => Promise<void>;
      readonly selectedItems: I[];
      removeSelectedItems?: unknown;
    }
  | {
      removeSelectedItems: () => Promise<void>;
      selectedItems?: unknown;
      removeItems?: unknown;
    }
) &
  (PreLoadStores extends true
    ? {
        loadAll: (selectedNamespaces: readonly string[]) => Promise<void>;
      }
    : {
        loadAll?: unknown;
      });

export type RenderHeaderTitle<Item extends ItemObject, PreLoadStores extends boolean> =
  | StrictReactNode
  | ((parent: NonInjectedItemListLayout<Item, PreLoadStores>) => StrictReactNode);

export type HeaderCustomizer = (placeholders: HeaderPlaceholders) => HeaderPlaceholders;
export type ItemListLayoutProps<Item extends ItemObject, PreLoadStores extends boolean = boolean> = {
  tableId?: string;
  className: IClassName;
  getItems: () => Item[];
  store: ItemListStore<Item, PreLoadStores>;
  dependentStores?: SubscribableStore[];
  preloadStores?: boolean;
  hideFilters?: boolean;
  searchFilters?: SearchFilter<Item>[];
  /** @deprecated */
  filterItems?: ItemsFilter<Item>[];

  // header (title, filtering, searching, etc.)
  showHeader?: boolean;
  headerClassName?: IClassName;
  renderHeaderTitle?: RenderHeaderTitle<Item, PreLoadStores>;
  customizeHeader?: HeaderCustomizer | HeaderCustomizer[];

  // items list configuration
  isReady?: boolean; // show loading indicator while not ready
  isSelectable?: boolean; // show checkbox in rows for selecting items
  isConfigurable?: boolean;
  copyClassNameFromHeadCells?: boolean;
  sortingCallbacks?: TableSortCallbacks<Item>;
  tableProps?: Partial<TableProps<Item>>; // low-level table configuration
  renderTableHeader?: (TableCellProps | undefined | null)[];
  renderTableContents: (item: Item) => (StrictReactNode | TableCellProps)[];
  renderItemMenu?: (item: Item, store: ItemListStore<Item, PreLoadStores>) => StrictReactNode;
  customizeTableRowProps?: (item: Item) => Partial<TableRowProps<Item>>;
  addRemoveButtons?: Partial<AddRemoveButtonsProps>;
  virtual?: boolean;

  // item details view
  hasDetailsView?: boolean;
  detailsItem?: Item;
  onDetails?: (item: Item) => void;

  // other
  customizeRemoveDialog?: (selectedItems: Item[]) => Partial<ConfirmDialogParams>;
  renderFooter?: (parent: NonInjectedItemListLayout<Item, PreLoadStores>) => StrictReactNode;

  spinnerTestId?: string;

  /**
   * Message to display when a store failed to load
   *
   * @default "Failed to load items"
   */
  failedToLoadMessage?: StrictReactNode;

  filterCallbacks?: ItemsFilters<Item>;
  "data-testid"?: string;
} & (PreLoadStores extends true
  ? {
      preloadStores?: true;
    }
  : {
      preloadStores: false;
    });

const defaultProps: Partial<ItemListLayoutProps<ItemObject, true>> = {
  showHeader: true,
  isSelectable: true,
  isConfigurable: false,
  copyClassNameFromHeadCells: true,
  preloadStores: true,
  dependentStores: [],
  searchFilters: [],
  customizeHeader: [],
  filterItems: [],
  hasDetailsView: true,
  onDetails: noop,
  virtual: true,
  customizeTableRowProps: () => ({}),
  failedToLoadMessage: "Failed to load items",
};

export interface ItemListLayoutStorage {
  showFilters: boolean;
}

interface Dependencies {
  selectedFilterNamespaces: IComputedValue<string[]>;
  itemListLayoutStorage: StorageLayer<ItemListLayoutStorage>;
  pageFiltersStore: PageFiltersStore;
}

@observer
class NonInjectedItemListLayout<I extends ItemObject, PreLoadStores extends boolean> extends React.Component<
  ItemListLayoutProps<I, PreLoadStores> & Dependencies
> {
  static defaultProps = defaultProps as object;

  constructor(props: ItemListLayoutProps<I, PreLoadStores> & Dependencies) {
    super(props);
    makeObservable(this);
    autoBindReact(this);
  }

  async componentDidMount() {
    const { isConfigurable, tableId, preloadStores } = this.props;

    if (isConfigurable && !tableId) {
      throw new Error("[ItemListLayout]: configurable list require props.tableId to be specified");
    }

    if (preloadStores) {
      const { store, dependentStores = [] } = this.props;
      const stores = Array.from(new Set([store, ...dependentStores])) as ItemListStore<I, true>[];

      stores.forEach((store) => store.loadAll(this.props.selectedFilterNamespaces.get()));
    }
  }

  get showFilters(): boolean {
    return this.props.itemListLayoutStorage.get().showFilters;
  }

  set showFilters(showFilters: boolean) {
    this.props.itemListLayoutStorage.merge({ showFilters });
  }

  @computed get filters() {
    let { activeFilters } = this.props.pageFiltersStore;
    const { searchFilters = [] } = this.props;

    if (searchFilters.length === 0) {
      activeFilters = activeFilters.filter(({ type }) => type !== FilterType.SEARCH);
    }

    return activeFilters;
  }

  toggleFilters() {
    this.showFilters = !this.showFilters;
  }

  @computed get isReady() {
    return this.props.isReady ?? this.props.store.isLoaded;
  }

  renderFilters() {
    const { hideFilters } = this.props;
    const { isReady, filters } = this;

    if (!isReady || !filters.length || hideFilters || !this.showFilters) {
      return null;
    }

    return <PageFiltersList filters={filters} />;
  }

  private filterCallbacks: ItemsFilters<I> = {
    [FilterType.SEARCH]: (items) => {
      const { searchFilters = [] } = this.props;
      const search = this.props.pageFiltersStore.getValues(FilterType.SEARCH)[0] || "";

      if (search && searchFilters.length) {
        const searchTexts = [search].map(normalizeText);

        return items.filter((item) =>
          searchFilters.some((getTexts) =>
            [getTexts(item)]
              .flat()
              .map(normalizeText)
              .some((source) => searchTexts.some((search) => source.includes(search))),
          ),
        );
      }

      return items;
    },
  };

  @computed get items() {
    const filterGroups = groupBy(this.filters, ({ type }) => type);
    const filterItems: ItemsFilter<I>[] = [];

    for (const [type, filtersGroup] of Object.entries(filterGroups)) {
      const filterCallback = this.filterCallbacks[type] ?? this.props.filterCallbacks?.[type];

      if (filterCallback && filtersGroup.length > 0) {
        filterItems.push(filterCallback);
      }
    }

    const items = this.props.getItems();

    return applyFilters(filterItems.concat(this.props.filterItems ?? []), items);
  }

  render() {
    const { renderHeaderTitle, "data-testid": dataTestId } = this.props;

    return untracked(() => (
      <div className={cssNames("ItemListLayout flex column", this.props.className)} data-testid={dataTestId}>
        <ItemListLayoutHeader
          getItems={() => this.items}
          getFilters={() => this.filters}
          toggleFilters={this.toggleFilters}
          store={this.props.store}
          searchFilters={this.props.searchFilters}
          showHeader={this.props.showHeader}
          headerClassName={this.props.headerClassName}
          renderHeaderTitle={
            typeof renderHeaderTitle === "function" ? () => renderHeaderTitle(this) : renderHeaderTitle
          }
          customizeHeader={this.props.customizeHeader}
        />

        <ItemListLayoutFilters
          getIsReady={() => this.isReady}
          getFilters={() => this.filters}
          getFiltersAreShown={() => this.showFilters}
          hideFilters={this.props.hideFilters ?? false}
        />

        <ItemListLayoutContent<I, PreLoadStores>
          getItems={() => this.items}
          getFilters={() => this.filters}
          tableId={this.props.tableId}
          className={this.props.className}
          store={this.props.store}
          getIsReady={() => this.isReady}
          isSelectable={this.props.isSelectable}
          isConfigurable={this.props.isConfigurable}
          copyClassNameFromHeadCells={this.props.copyClassNameFromHeadCells}
          sortingCallbacks={this.props.sortingCallbacks}
          tableProps={this.props.tableProps}
          renderTableHeader={this.props.renderTableHeader}
          renderTableContents={this.props.renderTableContents}
          renderItemMenu={this.props.renderItemMenu}
          customizeTableRowProps={this.props.customizeTableRowProps}
          addRemoveButtons={this.props.addRemoveButtons}
          virtual={this.props.virtual}
          hasDetailsView={this.props.hasDetailsView}
          detailsItem={this.props.detailsItem}
          onDetails={this.props.onDetails}
          customizeRemoveDialog={this.props.customizeRemoveDialog}
          failedToLoadMessage={this.props.failedToLoadMessage}
          spinnerTestId={this.props.spinnerTestId}
        />

        {this.props.renderFooter?.(this)}
      </div>
    ));
  }
}

export const ItemListLayout = withInjectables<Dependencies, ItemListLayoutProps<ItemObject, boolean>>(
  NonInjectedItemListLayout,
  {
    getProps: (di, props) => ({
      ...props,
      selectedFilterNamespaces: di.inject(selectedFilterNamespacesInjectable),
      itemListLayoutStorage: di.inject(itemListLayoutStorageInjectable),
      pageFiltersStore: di.inject(pageFiltersStoreInjectable),
    }),
  },
) as <I extends ItemObject, PreLoadStores extends boolean = true>(
  props: ItemListLayoutProps<I, PreLoadStores>,
) => React.ReactElement;

function applyFilters<I extends ItemObject>(filters: ItemsFilter<I>[], items: I[]): I[] {
  if (!filters || !filters.length) {
    return items;
  }

  return filters.reduce((items, filter) => filter(items), items);
}
