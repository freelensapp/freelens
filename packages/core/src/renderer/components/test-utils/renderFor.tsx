/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = testingLibraryRender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>);

  return {
    ...result,

    rerender: (ui: React.ReactElement) => result.rerender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>),
  };
};
