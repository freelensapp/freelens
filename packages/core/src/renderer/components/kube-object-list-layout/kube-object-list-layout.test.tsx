/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import "@testing-library/jest-dom";
import { podListLayoutColumnInjectionToken } from "@freelensapp/list-layout";
import type { RenderResult } from "@testing-library/react";
import React from "react";
import appPathsStateInjectable from "../../../common/app-paths/app-paths-state.injectable";
import directoryForKubeConfigsInjectable from "../../../common/app-paths/directory-for-kube-configs/directory-for-kube-configs.injectable";
import directoryForUserDataInjectable from "../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import { Cluster } from "../../../common/cluster/cluster";
import isTableColumnHiddenInjectable from "../../../features/user-preferences/common/is-table-column-hidden.injectable";
import hostedClusterInjectable from "../../cluster-frame-context/hosted-cluster.injectable";
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import storesAndApisCanBeCreatedInjectable from "../../stores-apis-can-be-created.injectable";
import kubeSelectedUrlParamInjectable from "../kube-detail-params/kube-selected-url.injectable";
import toggleKubeDetailsPaneInjectable from "../kube-detail-params/toggle-details.injectable";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import type { PodStore } from "../workloads-pods/store";
import podStoreInjectable from "../workloads-pods/store.injectable";
import { KubeObjectListLayout } from "./index";

describe("kube-object-list-layout", () => {
  let di: DiContainer;
  let render: DiRender;
  let podStore: PodStore;

  beforeEach(() => {
    di = getDiForUnitTesting();

    di.override(directoryForUserDataInjectable, () => "/some-user-store-path");
    di.override(directoryForKubeConfigsInjectable, () => "/some-kube-configs");
    di.override(storesAndApisCanBeCreatedInjectable, () => true);
    di.override(isTableColumnHiddenInjectable, () => () => false);

    di.override(
      hostedClusterInjectable,
      () =>
        new Cluster({
          contextName: "some-context-name",
          id: "some-cluster-id",
          kubeConfigPath: "/some-path-to-a-kubeconfig",
        }),
    );

    render = renderFor(di);

    di.override(subscribeStoresInjectable, () => jest.fn().mockImplementation(() => jest.fn()));
    di.override(kubeSelectedUrlParamInjectable, () => ({
      get: () => "path",
    }));
    di.override(toggleKubeDetailsPaneInjectable, () => null);
    di.override(appPathsStateInjectable, () => ({
      get: () => ({}),
    }));

    podStore = di.inject(podStoreInjectable);
  });

  describe("given pod store", () => {
    let result: RenderResult;

    it("renders", () => {
      result = render(
        <div>
          <KubeObjectListLayout
            className="Pods"
            store={podStore}
            tableId="workloads_pods"
            isConfigurable
            renderHeaderTitle="Pods"
            renderTableContents={(pod) => [<div key={pod.getName()}>{pod.getName()}</div>]}
            columns={di.injectMany(podListLayoutColumnInjectionToken)}
          />
        </div>,
      );

      expect(result.baseElement).toMatchSnapshot();
    });

    describe("given resourcename", () => {
      it("uses resourcename in search placeholder", () => {
        result = render(
          <div>
            <KubeObjectListLayout
              className="Pods"
              store={podStore}
              tableId="workloads_pods"
              isConfigurable
              renderHeaderTitle="Pods"
              renderTableContents={(pod) => [<div key={pod.getName()}>{pod.getName()}</div>]}
              resourceName="My Custom Items"
              searchFilters={[() => null]}
            />
          </div>,
        );

        expect(result.getByPlaceholderText("Search My Custom Items...")).toBeInTheDocument();
      });
    });
  });
});
