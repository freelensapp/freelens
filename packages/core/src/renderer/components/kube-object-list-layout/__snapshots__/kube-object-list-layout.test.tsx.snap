// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kube-object-list-layout given pod store renders 1`] = `
<body>
  <div>
    <div>
      <div
        class="ItemListLayout flex column KubeObjectListLayout Pods"
      >
        <div
          class="header flex gaps align-center"
        >
          <h5
            class="title"
          >
            Pods
          </h5>
          <div
            class="info-panel box grow"
          >
            0 items
          </div>
          <div
            class="NamespaceSelectFilterParent"
            data-testid="namespace-select-filter"
          >
            <div
              class="Select theme-dark NamespaceSelect NamespaceSelectFilter css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-kube-object-list-layout-namespace-select-input-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class="Select__control css-13cymwt-control"
              >
                <div
                  class="Select__value-container Select__value-container--is-multi css-hlgwow"
                >
                  <div
                    class="Select__placeholder css-1jqq78o-placeholder"
                    id="react-select-kube-object-list-layout-namespace-select-input-placeholder"
                  >
                    All namespaces
                  </div>
                  <div
                    class="Select__input-container css-19bb58m"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-kube-object-list-layout-namespace-select-input-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class="Select__input"
                      id="kube-object-list-layout-namespace-select-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="Select__indicators css-1wy0on6"
                >
                  <span
                    class="Select__indicator-separator css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="Select__indicator Select__dropdown-indicator css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="items box grow flex column"
        >
          <div
            class="Table flex column KubeObjectListLayout Pods box grow dark selectable scrollable sortable autoSize virtual"
          >
            <div
              class="TableHead sticky nowrap topLine"
            >
              <div
                class="TableCell checkbox"
              >
                <label
                  class="Checkbox flex align-center"
                >
                  <input
                    type="checkbox"
                  />
                  <i
                    class="box flex align-center"
                  />
                </label>
              </div>
              <div
                class="TableCell name nowrap sorting"
                id="name"
              >
                <div
                  class="content"
                >
                  Name
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell warning nowrap"
              >
                <div
                  class="content"
                />
              </div>
              <div
                class="TableCell namespace nowrap sorting"
                id="namespace"
              >
                <div
                  class="content"
                >
                  Namespace
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell containers nowrap sorting"
                id="containers"
              >
                <div
                  class="content"
                >
                  Containers
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell restarts nowrap sorting"
                id="restarts"
              >
                <div
                  class="content"
                >
                  Restarts
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell owners nowrap sorting"
                id="owners"
              >
                <div
                  class="content"
                >
                  Controlled By
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell node nowrap sorting"
                id="node"
              >
                <div
                  class="content"
                >
                  Node
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell qos nowrap sorting"
                id="qos"
              >
                <div
                  class="content"
                >
                  QoS
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell age nowrap sorting"
                id="age"
              >
                <div
                  class="content"
                >
                  Age
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell status nowrap sorting"
                id="status"
              >
                <div
                  class="content"
                >
                  Status
                </div>
                <i
                  class="Icon sortIcon material focusable"
                >
                  <span
                    class="icon"
                    data-icon-name="arrow_drop_down"
                  >
                    arrow_drop_down
                  </span>
                </i>
              </div>
              <div
                class="TableCell menu nowrap"
              >
                <div
                  class="content"
                >
                  <i
                    class="Icon material interactive focusable"
                    id="menu-actions-for-item-object-list-content"
                    tabindex="0"
                  >
                    <span
                      class="icon"
                      data-icon-name="more_vert"
                    >
                      more_vert
                    </span>
                  </i>
                </div>
              </div>
            </div>
            <div
              class="Spinner singleColor center"
              data-testid="kube-object-list-layout-spinner"
            />
          </div>
          <div
            class="AddRemoveButtons flex gaps"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
